core/src/consensus/tendermint/mod.rs

```diff
@@ -726,35 +720,53 @@ impl TendermintInner {
         let precommit_hash = message_hash(step, *header.parent_hash());
         let precommits_field =
             &header.seal().get(2).expect("block went through verify_block_basic; block has .seal_fields() fields; qed");
-        let mut origins = HashSet::new();
+        let mut precommit_bitset: BitSet = UntrustedRlp::new(
+            &header.seal().get(3).expect("block went through verify_block_basic; block has .seal_fields() fields; qed"),
+        )
+        .as_val()?;
+        let mut bitset_index = 0;
+        let mut counter = 0;
         for rlp in UntrustedRlp::new(precommits_field).iter() {
+            if precommit_bitset.is_empty() {
+                cwarn!(
+                    ENGINE,
+                    "verify_block_external: Precommit bitset got empty while dealing with validators of block hash {} on the seal.",
+                    header.parent_hash()
+                );
+                return Err(BlockError::InvalidSeal.into())
+            }
+            while !precommit_bitset.is_set(bitset_index) {
+                bitset_index += 1;
+            }
+            precommit_bitset.reset(bitset_index);
+
+            // This message is a dummy message to check address. signer_index value does not affect.
             let precommit = ConsensusMessage {
```

Comment:
It seems that only signature is used.
How about replacing it to `let signature = rlp.as_val();`

Reply:
If type inference doesn't work,
```rust
let signature: SchnorrSignature = rlp.as_val()?;
```

Reply:
I forgot to remove this. I'll fix this. Type inference should work as I think.

Reply:
Fixed.